contractcourt/htlc_outgoing_contest_resolver.go

```diff
@@ -55,38 +55,74 @@ func (h *htlcOutgoingContestResolver) Resolve() (ContractResolver, error) {
 		return nil, nil
 	}
 
+	// handleSpendEvent is a helper closure that processes the spend event
+	// received.
+	handleSpendEvent := func(event *spendResult) (ContractResolver, error) {
+		// Return the error if there's one.
+		if event.err != nil {
+			return nil, event.err
+		}
+
+		// If an event is returned but the spend is nil, then the HTLC
+		// has been spent in a timeout tx. In this case, we'll let the
+		// timeout resolver to handle it.
+		if event.spend == nil {
+			log.Infof("HTLC(%s) has already been spent in "+
+				"timeout tx", h.HtlcPoint())
+			return nil, nil
+		}
+
+		// The only way this output can be spent by the remote party is
+		// by revealing the preimage. So we'll perform our duties to
+		// clean up the contract once it has been claimed.
+		return h.claimCleanUp(event.spend)
+	}
+
 	// Otherwise, we'll watch for two external signals to decide if we'll
 	// morph into another resolver, or fully resolve the contract.
 	//
-	// The output we'll be watching for is the *direct* spend from the HTLC
-	// output. If this isn't our commitment transaction, it'll be right on
-	// the resolution. Otherwise, we fetch this pointer from the input of
-	// the time out transaction.
-	outPointToWatch, scriptToWatch, err := h.chainDetailsToWatch()
-	if err != nil {
-		return nil, err
-	}
+	// First, we'll start watching for the spend of the HTLC. This is
+	// either a spend found in mempool which reveals the preimage, or a
+	// spend found in blocks that is a timeout.
+	//
+	// NOTE: We expect the HTLC to be spent in a timeout tx if it's found
+	// in a block since for the preimage spend, we'd already catch it in
+	// our mempool. This could be false if our mempool missed the preimage
+	// spend tx. In that case, we'd still see the preimage spend tx in
+	// blocks.
+	resultChan := make(chan *spendResult, 1)
+	go func() {
+		sr := &spendResult{}
+
+		// Watch the spend in a goroutine and send back the result.
+		result, err := h.watchHtlcSpend()
+		if err != nil {
+			// Send back the err if there's one.
+			sr.err = err
+
+		} else if isPreimageSpend(result, h.isLocalCommit()) {
+			// Otherwise, we want only to send back the spend that
+			// reveals the preimage.
+			//
+			// NOTE: this could be a timeout tx if there's a race
+			// between the contest resolver and the timeout
+			// resolver.
+			sr.spend = result
+		}
 
-	// First, we'll register for a spend notification for this output. If
-	// the remote party sweeps with the pre-image, we'll be notified.
-	spendNtfn, err := h.Notifier.RegisterSpendNtfn(
```

Comment:
Why not inherit the approach of the normal timeout resolver that watches _both_ signals? Given the context embeds the timeout, we could leverage methods on the timeout reoslver to reuse more code.

Reply:
Yeah this is done via the method `h.watchHtlcSpend`, which watches the two signals, and that method is in htlc timeout resolver.

Here we only call that method and return that spend details in a channel because in htlc outgoing contest resolver we need to watch three signals - the above two, plus block subscription.